Add plugin command to provide for users how find and install / uninstall

horike37 opened the pull request:
<!--
1. Please check out and follow our Contributing Guidelines: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md
2. Do not remove any section of the template. If something is not applicable leave it empty but leave it in the PR
3. Please follow the template, otherwise we'll have to ask you to update it and it will take longer until your PR is merged
-->

## What did you implement:
Re-work for #2861 and #2907

Added top level commands to the framework core.
`serverless plugin`

- `serverless plugin` - shows list of sub commands
- `serverless plugin install` - downloads plugin and automatically adds to serverless.yml.
- `serverless plugin uninstall` - removes the plugin from serverless.yml file
- `serverless plugin list` - list all available/known plugins from the [plugin repo](https://github.com/serverless/plugins)
- `serverless plugin search` - search for match out of list.

<!--
Briefly describe the feature if no issue exists for this PR
-->

## How did you implement it:
I have restored the code from #2907 and applied some slightly updates, but the basic spec has not changed
<!--
If this is a nontrivial change please briefly describe your implementation so its easy for us to understand and review your code.
-->

## How can we verify it:
Please run the following commands
- serverless plugin list
- serverless plugin search --query / -q {query string}
- serverless plugin install --name / -n {plugin name}
- serverless plugin uninstall --name / -n {plugin name}

<!--
Add any applicable config, commands, screenshots or other resources
to make it easy for us to verify this works. The easier you make it for us
to review a PR, the faster we can review and merge it.

Examples:
* serverless.yml - Fully functioning to easily deploy changes
* Screenshots - Showing the difference between your output and the master
* Cloud Configuration - List cloud resources and show that the correct configuration is in place (e.g. AWS CLI commands)
* Other - Anything else that comes to mind to help us evaluate
-->

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

DavidWells commented:
Looks good. I submitted a couple updates to the `plugin list` output

@horike37 seeing an issue with uninstall: https://www.screencast.com/t/HlAGaRpt14t

```
myService git:(add-plugin-command) ✗ ../bin/serverless plugin uninstall --name serverless-webpack

  Serverless Error ---------------------------------------

  Serverless plugin "serverless-webpack" not found. Make sure it's installed and listed in the "plugins" section of your serverless config file.

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless

  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           6.6.0
     Serverless Version:     1.19.0
```

For some reason this error https://github.com/serverless/serverless/blob/master/lib/classes/PluginManager.js#L70 is throwing from inside the test service.

horike37 commented:
@DavidWells 
Thank you for checking the behavior!
I have reproduced that. Seems that the problem depends on `serverless-webpack`.
Because the others plugins has not faced the same problem, but only when `serverless-webpack` is installed, all subcommand don't work...

In addition, I have tried to manually install ( Run `npm install serverless-webpack --save-dev` and adding `- serverless-webpack` into `plugins` section ) but I have faced the same problem.

Hey @HyperBrain , any ideas why the following error message with `serverless-webpack`?
https://www.screencast.com/t/HlAGaRpt14t

HyperBrain commented:
@horike37 It seems that the webpack-plugin crashes on one of the operations. The error message `Serverless plugin "serverless-webpack" not found.` is misleading here -> this will be shown as soon as the loadPlugin() for a plugin leads to a crash. It captures any exception/rejection and prints out this error message.
We should improve that in the framework, so that it is possible to find the exact callstack, i.e. the catch in pluginManager should expose the original callstack and not create a new error object.

Is there any new requirement for defining plugins with the new command/subcommands? Do the plugins have to have some new methods?

HyperBrain commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,343 @@
+            .log(`Installing plugin "${pluginName}" (this might take a few seconds...)`);
+
+          childProcess
+            .execSync(`npm install --save-dev ${pluginName}`, {
```
Getting the latest version of the plugin might not be what the user wants. There must be a "version" option to specify the version that has to be installed (e.g. npm version pattern).
```
.execSync(`npm install --save-dev ${pluginName}@${pluginVersion}`, {
```

HyperBrain commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,343 @@
+
+  getServerlessFilePath() {
+    const servicePath = this.serverless.config.servicePath;
+    const serverlessYmlFilePath = path.join(servicePath, 'serverless.yml');
```
The serverless file can also be a JSON file => serverless.yaml, serverless.yml and serverless.json should be searched here

HyperBrain commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,343 @@
+              // replace the plugins definition in the serverless file
+              serverlessFileContent = serverlessFileContent
+                .replace(
+                  /(plugins:)+(\s)*(- .*\s*)*(\[\])*/, // eslint-disable-line no-useless-escape
```
This will not work if a user has commented out a line in the plugins array with "#" which is completely valid.

DavidWells replied:
@horike37 we need to use a YAML ast parser. The framework in general needs to use an AST parser to avoid regex issues.

see https://github.com/serverless/dashboard/blob/master/package.json#L125

& https://github.com/serverless/dashboard/blob/master/app/utils/yaml/updateYamlContents.js

Lets get the YAML AST parse into framework core and as a /util function that anything can use

HyperBrain commented:
@horike37 I will check the uninstall and try to get the original callstack from within the plugin. It might have something to do with the initialization of the plugin - the "plugin uninstall" command runs within the serverless context, i.e. it first loads all plugins that are installed (through pluginManager). It could be that the webpack plugin bails out because some configuration is missing.
With the callstack I can fix that easily if it's the case.

pmuens commented on docs/providers/aws/README.md:
```diff
@@ -68,6 +68,10 @@ If you have questions, join the [chat in gitter](https://gitter.im/serverless/se
+        <li><a href="./cli-reference/plugin-list.md">Plugin List</a></li>
+        <li><a href="./cli-reference/plugin-search.md">Plugin Search</a></li>
+        <li><a href="./cli-reference/plugin-install.md">Plugin Install</a></li>
+        <li><a href="./cli-reference/plugin-uninstall.md">Plugin Uninstall</a></li>
```
All those docs should be added to the other provider documentations as well.

brianneisler commented:
@horike37 how much work would it be to switch the order of commands here? 

`serverless plugin install`       -> `serverless install plugin`
`serverless plugin uninstall`   -> `serverless uninstall plugin`
`serverless plugin list`            -> `serverless list plugin`
`serverless plugin search`      -> `serverless search plugin`

I ask because the order of these commands feels very inconsistent with the way that we currently handle functions and credentials. (basically the usual order is `serverless action object`)

Examples
`serverless config`
`serverless config credentials`
`serverless deploy`
`serverless deploy function`
`serverless rollback`
`serverless rollback function`

HyperBrain commented:
@brianneisler This would mean to have separate commands `install`, `uninstall`, `list` and `search`. Right now commands are groupings of functionality. I'm not sure if this change in semantics wouldn't lead to more confusion in the end.

@pmuens @horike37 @eahefnawy What do you think?

DavidWells commented:
I agree with @HyperBrain we should not switch the order of commands as @brianneisler describes. 

It will leads to pollution of top level commands (aka `list`), that really only serve one purpose. Same with search. 

I can see `sls install plugin` being nice, but we could alias that command to `sls plugin install` instead.

In the future, *if* we have a `sls search` command we can alias the command to `sls plugin search`

horike37 commented:
@HyperBrain 
Thank you for the taking a look for the trouble and reviwing! 

>Is there any new requirement for defining plugins with the new command/subcommands? Do the plugins have to have some new methods?

There is not an aditional requirement/config for defining plugins with the command. Those remain as it was.

>I will check the uninstall and try to get the original callstack from within the plugin.

thx! I will do something debugging for this too.

@brianneisler @HyperBrain @DavidWells @pmuens 
That's make sense. 
My personal preference is `serverless plugin <action>` order, since I feel that the order would be easier to understand the meaning of the command than `serverless action object` order . 

We certainly should follow unified rules to avoid confusion, but in this case I do not think it will confuse users.

pmuens commented:
🤔 that's a tough one. I'm also in favor to settle on some consistency here.

`serverless install plugin` reads very nicely and makes sense.

It's also great since we already have an `install` command so this could be a "natural" extension. However if we go down that path we should update the `install` command to smth. like `serverless install service` to avoid confusion. So this could also be a downside since we have to take care of backwards compatibility if we change something here.

I like the grouping you get with `serverless plugin <action>` since it follows a predictable schema. This way I always know what other commands I could use (and maybe even try without knowing).

brianneisler commented:
Ok, seems like the consensus here is to go with `serverless plugin <action>`. I'm fine with that. 

Seems like we should introduce a similar command for functions then to help keep things consistent. 
```
serverless function <action>
```

We can keep the `serverless deploy function` and `serverless rollback function` commands in v1 to keep from breaking things and then deprecate them in v2.

pmuens commented:
Sorry for crashing the party here 😬

After thinking about this again over the last couple of days I'm in the `serverless install` camp again.

The reason for that is that we should get beyond installing just plugins quite soon.

Users should be able to extend their services by `install`ing individual functions, other services, plugins, toolings and much more in the future.

That's because I think grouping everything under `serverless install` might be way easier to remember and deal with as the support for installing 3rd party content grows.

horike37 commented:
We have a nice discussion here :smile:

If the command order issue is limited to `install` I agree with @pmuens. 
I think `install` is a generic action so we can get extensibility by adding the top level commands

However, most of the actions are not generic. For instance, if we want to add the non-generic action something like `register` or `visualize` to the top level, we can not get the extensibility. I feel that such as `sls register` or `sls visualize` is weird.
- `sls plugin register` - registers a plugin in the plugin repository
- `sls plugin visualize` https://github.com/serverless/serverless/pull/2821

From the viewpoint of getting more extensibility in the future, I prefer `sls plugin <action>`.

pmuens commented:
Thanks for the feedback @horike37 👍

Good point. What about having both?

So for the `install`, `uninstall`, `search` etc. functionality (for 3rd party extensions / content) we have the `serverless install`, `serverless uninstall`, ... commands. Whereas for the non-generic actions we simply add them like you've described above (e.g. `serverless plugin visualize`).

DavidWells commented:
> I can see sls install plugin being nice, but we could alias that command to sls plugin install instead.

I like this approach @pmuens 

Both works for me.

- serverless plugin - shows list of sub commands
- serverless plugin install - downloads plugin and automatically adds to serverless.yml.
- serverless install plugin - alias to serverless plugin install
- serverless plugin uninstall - removes the plugin from serverless.yml file
- serverless plugin list - list all available/known plugins from the plugin repo
- serverless plugin search - search for match out of list.

I am not sure how `sls search` or `sls uninstall` would work. I think we should not implement those at this time

horike37 commented:
I like @pmuens' solution too.

How about adding `search` to the top level when we need to use search other than `plugin` in the future?

pmuens commented:
Yep, that sounds like a good solution @DavidWells 👍

@HyperBrain already wrote a nice proposal for plugin alias support in https://github.com/serverless/serverless/issues/4135 💪

horike37 commented:
OK. will do :muscle:

horike37 commented:
As I commented above, after installing serverless-webpack, all commands are broken.
I investigated the cause and found that serverless-webpack plugin needs to install a webpack module manually. Is it correct? @HyperBrain

Right now, if SLS_DEBUG set, an original exception has become to be displayed as an error message by HyperBrain'PR(https://github.com/serverless/serverless/commit/bb1e74a43cb91e254be216f8a3d0aa806060fa53). However, I guess that users can't understand what's happening to see only an exception message.

```
$../bin/serverless info
 
  Error --------------------------------------------------
 
  Cannot find module 'webpack'
 
     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
 
  Stack Trace --------------------------------------------
 
Error: Cannot find module 'webpack'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/horike/src/serverless/tmp/node_modules/serverless-webpack/lib/compile.js:4:17)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)

```

I think had better be able to see an original exception even w/o SLS_DEBUG and improve an error message something like this to be more user-friendly

```
Error --------------------------------------------------

`${error.message}. It seems that the installed plugin is broken or lacks some additional configurations or operations for the plugin. Make sure those.`

  Stack Trace --------------------------------------------

```
@pmuens @DavidWells @brianneisler @HyperBrain 
What do you think?

HyperBrain commented:
@horike37 You're right. Webpack is a peer dependency needed by serverless-webpack. You should see the peer dependency warning already if you do a npm install in your project.
Just add `webpack@^3.5.0` as dependency to your project and it should work. Of course you also can use older versions - that's your choice.

HyperBrain commented:
BTW: If the plugin install command adds and installs plugins, it should also evaluate the needed peer dependencies and install them too (by default as the latest version).

pmuens commented:
> BTW: If the plugin install command adds and installs plugins, it should also evaluate the needed peer dependencies and install them too (by default as the latest version).

> There is another thing: Most plugins will need some basic configuration to make them usable. I'm thinking here of something like our template creation command "serverless create". Plugins could contain some default config template that gets installed when you install the plugin with the command. The plugins should therefore expose a new static method install() that will be called by plugin install. Then the plugin would create a minimal working configuration, if one does not yet exist. If the static method is missing (for old plugins), sls plugin install just would skip it.

> We can do something like this with the webpack plugin if you like, for testing. I'm open to prepare the feature plugin-side and release an unofficial beta on npm.

I like that idea by @HyperBrain 👍.

I'm just wondering if it's too much hassle for the plugin developer to write this extra logic. Additionally we currently have a couple of old plugins which needs to be updated so that the setup scripts are included.

My proposal would be to display a link to the GitHub / Plugins `README.md` after installing the plugin (since this always includes the steps on how to use it). Something like:

```bash
Successfully installed serverleqss-webpack v3.0.0

For setup instructions see:
https://github.com/elastic-coders/serverless-webpack/blob/master/README.md
```

We could grab that from the `npm` package or add a separate property to our `plugins.json` file in our `plugins` GitHub repo.

We could also print the `,md` file content, but IMHO that could result in too much text in the console.

horike37 commented:
>You're right. Webpack is a peer dependency needed by serverless-webpack. You should see the peer dependency warning already if you do a npm install in your project. This allows you to choose which webpack version you want to use

That makes sense!!!! I didn't know that. I will do so.
Thank you for the advice  😄

Then, I like @HyperBrain' idea too.

>My proposal would be to display a link to the GitHub / Plugins README.md after installing the plugin (since this always includes the steps on how to use it).

Good idea @pmuens 👍
I like the idea of adding a separate property to our plugins.json file since it can be displayed intentionally what plugin developers want.

horike37 commented:
@HyperBrain 
Thank you for commenting about peer dependency :smile:
I have implemented as follow on the d5f2d02 commit.
1. Detected peerDependencies syntax in package.json of the installed plugin
2. Added the peerDependencies syntax to devDependencies with package.json of the service
3. Run `npm install` on the service directory

Therefore peerDependencies lib would be installed with the version range which is described in package.json of the installed plugin, so I guess the range check don't need. Is this understanding mistaken?

HyperBrain commented:
@horike37 Yes, I think the explicit check is not needed then, because npm itself uses `semver` internally and should handle it correctly. Maybe a final test with a plugin that has a ranged peer dep (if there is any plugin at all?) whould prove that.

horike37 commented:
@HyperBrain 
Got it! thx.

@pmuens @DavidWells
I have fixed all the things which were pointed out. :bowtie:
Please review again.

** The list of what I have done after the first review
- Added the feature to all providers document
- Also applied this feature to serverless.json
- Added yamlAstParser.js as utils library so that it makes easy to operate yaml files
- Added `--version` option to the install command
- Applied the command order of the final spec
- Also updated that peerDependencies are installed automatically when run `sls plugin install` or `sls install plugin`

brianneisler commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,415 @@
+            `Installing plugin "${pluginName}@${pluginVersion}" (this might take a few seconds...)`
+          );
+
+          childProcess
```
@horike37 please avoid doing these kinds of things synchronously. If you need to wait on something to complete it is better to use a Promise and make it part of the async chain.

brianneisler commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,415 @@
+            fs.readFileSync(packageJsonFilePath).toString()
+          ).devDependencies[pluginName];
+          if (pluginInstalled) {
+            this.addPluginToServerlessFile(pluginName);
```
Need to validate that the same plugin is not being installed twice

brianneisler commented on lib/utils/yamlAstParser.js:
```diff
@@ -0,0 +1,108 @@
+const _ = require('lodash');
+const os = require('os');
+
+const astObject = {};
```
@horike37 putting this here is a very dangerous pattern. This will immediately break if the util methods are used more than once in the same execution lifecycle or if there are multiple run in parallel since all calls are sharing the same object in memory. Instead the AST should be passed as a function parameter.

horike37 commented:
Thank you for reviewing @brianneisler.
Just updated based on your feedback.

horike37 commented:
I just confirmed #4172 and @HyperBrain's approach on #4135.
I already added aliases of the install and uninstall on this PR like below by adding a new lifecycle event, but it would be a bad way.
Therefore, I should remove one of `plugin <action>` or `<action> plugin` from this PR once. Then, will apply the aliases system to the plugin command again after adding it.

https://github.com/horike37/serverless/blob/34d474fb96a0705195246765adde6eb25181e50f/lib/plugins/plugin/plugin.js#L24-L28
https://github.com/horike37/serverless/blob/34d474fb96a0705195246765adde6eb25181e50f/lib/plugins/plugin/plugin.js#L75-L81

pmuens commented on docs/providers/openwhisk/cli-reference/plugin-install.md:
```diff
@@ -0,0 +1,35 @@
+-->
+
+<!-- DOCS-SITE-LINK:START automatically generated  -->
+### [Read this on the main serverless docs site](https://www.serverless.com/framework/docs/providers/aws/cli-reference/plugin-install)
```
This needs to be `openwhisk`

pmuens commented on docs/providers/openwhisk/cli-reference/plugin-list.md:
```diff
@@ -0,0 +1,25 @@
+-->
+
+<!-- DOCS-SITE-LINK:START automatically generated  -->
+### [Read this on the main serverless docs site](https://www.serverless.com/framework/docs/providers/aws/cli-reference/plugin-list)
```
This needs to be `openwhisk`

pmuens commented on docs/providers/openwhisk/cli-reference/plugin-search.md:
```diff
@@ -0,0 +1,33 @@
+-->
+
+<!-- DOCS-SITE-LINK:START automatically generated  -->
+### [Read this on the main serverless docs site](https://www.serverless.com/framework/docs/providers/aws/cli-reference/plugin-search)
```
This needs to be `openwhisk`

pmuens commented on docs/providers/openwhisk/cli-reference/plugin-uninstall.md:
```diff
@@ -0,0 +1,33 @@
+-->
+
+<!-- DOCS-SITE-LINK:START automatically generated  -->
+### [Read this on the main serverless docs site](https://www.serverless.com/framework/docs/providers/aws/cli-reference/plugin-uninstall)
```
This needs to be `openwhisk`

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+
```
This newline should be removed.

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+    return BbPromise.resolve();
+  }
+
+  getServerlessFilePath() {
```
This could be replace with this util function: https://github.com/serverless/serverless/blob/master/lib/utils/getServerlessConfigFile.js

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+    const pluginVersion = this.options.version || 'latest';
+
+    // check if package.json is already present. Otherwise create one
+    if (!fs.existsSync(packageJsonFilePath)) {
```
Could we make this check async?

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+        devDependencies: {},
+      };
+
+      fse.writeJsonSync(packageJsonFilePath, packageJsonFileContent);
```
Could we make this async?

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+    );
+
+    childProcess
+      .execSync(`npm install --save-dev ${pluginName}@${pluginVersion}`, {
```
Could we make this async?

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+        stdio: 'ignore',
+      });
+    const pluginInstalled = !!JSON.parse(
+      fs.readFileSync(packageJsonFilePath).toString()
```
Could we make this call async?

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+    const servicePath = this.serverless.config.servicePath;
+    const packageJsonFilePath = path.join(servicePath, 'package.json');
+    childProcess
+      .execSync(`npm uninstall --save-dev ${pluginName}`, {
```
Could we turn this into an async call?

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+
+    // check if plugin was uninstalled correctly
+    const pluginStillAvailable = !!JSON.parse(
+      fs.readFileSync(packageJsonFilePath).toString()
```
Could this be turned into an async call?

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+    const pluginName = this.options.name;
+    const serverlessFilePath = this.getServerlessFilePath();
+    if (_.last(_.split(serverlessFilePath, '.')) === 'json') {
+      const serverlessFileObj = fse.readJsonSync(serverlessFilePath);
```
Could this be turned into an async call?

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+        serverlessFileObj.plugins = [pluginName];
+      }
+      serverlessFileObj.plugins = _.sortedUniq(serverlessFileObj.plugins);
+      fse.writeJsonSync(serverlessFilePath, serverlessFileObj);
```
Could this be turned into an `async` call?

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+          _.unset(serverlessFileObj, 'plugins');
+        }
+      }
+      fse.writeJsonSync(serverlessFilePath, serverlessFileObj);
```
This should be turned into an async call.

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+    const pluginName = this.options.name;
+    const serverlessFilePath = this.getServerlessFilePath();
+    if (_.last(_.split(serverlessFilePath, '.')) === 'json') {
+      const serverlessFileObj = fse.readJsonSync(serverlessFilePath);
```
This should be turned into an async call.

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+    const servicePackageJsonFilePath = path.join(this.serverless.config.servicePath,
+      'package.json');
+    if (fs.existsSync(pluginPackageJsonFilePath) && fs.existsSync(servicePackageJsonFilePath)) {
+      const pluginPackageJson = fse.readJsonSync(pluginPackageJsonFilePath);
```
This should be turned into an async call.

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+    if (fs.existsSync(pluginPackageJsonFilePath) && fs.existsSync(servicePackageJsonFilePath)) {
+      const pluginPackageJson = fse.readJsonSync(pluginPackageJsonFilePath);
+      if (pluginPackageJson.peerDependencies) {
+        const servicePackageJson = fse.readJsonSync(servicePackageJsonFilePath);
```
This should be turned into an async call.

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+        const servicePackageJson = fse.readJsonSync(servicePackageJsonFilePath);
+        servicePackageJson.devDependencies =
+          _.merge(servicePackageJson.devDependencies, pluginPackageJson.peerDependencies);
+        fse.writeJsonSync(servicePackageJsonFilePath, servicePackageJson);
```
This should be turned into an async call.

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+          _.merge(servicePackageJson.devDependencies, pluginPackageJson.peerDependencies);
+        fse.writeJsonSync(servicePackageJsonFilePath, servicePackageJson);
+        childProcess
+          .execSync('npm install', {
```
This should be turned into an async call.

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+      if (pluginPackageJson.peerDependencies) {
+        _.forEach(pluginPackageJson.peerDependencies, (v, k) => {
+          childProcess
+            .execSync(`npm uninstall --save-dev ${k}`, {
```
This should be turned into an async call.

pmuens commented on lib/plugins/plugin/plugin.js:
```diff
@@ -0,0 +1,459 @@
+    const pluginName = this.options.name;
+    const pluginPackageJsonFilePath = path.join(this.serverless.config.servicePath,
+      'node_modules', pluginName, 'package.json');
+    if (fs.existsSync(pluginPackageJsonFilePath)) {
```
this should be turned into an async call.

HyperBrain commented:
@pmuens @horike37 If I'll be able to spend some time on it, I can take care of the command alias implementation in a separate PR, and command aliases could be added afterwards, for the plugin command too.

eahefnawy commented on docs/providers/aws/cli-reference/install-plugin.md:
```diff
@@ -0,0 +1,36 @@
+
+Install a Serverless plugin and add it to the services `plugins` array.
+
+**Note:** You might want to change the order of the plugin in the services `plugins` array.
```
lol why?!

pmuens replied:
Because plugins are loaded in the order you list them in the `plugins` array.

I you e.g. want to overwrite a functionality from another plugin you'd have to put this plugin after the one you want to overwrite.

horike37 commented:
@pmuens @eahefnawy 
Thank you for reviewing and testing on this!

>We're currently enforcing the policy that all new code should be as async as possible since sync calls are on of the roots of the sloweness of the Framework.

That sounds reasonable :star2: I will update those.

>Could you also rebase (or merge master into) this branch and add the corresponding docs for the recently added kubeless provider?
>So plugin <action> is sufficient. The alias support can be added later on with the help of the proposal @HyperBrain recently wrote (sorry for the confusion about that 😞). Please don't forget to remove the docs for the alias support as well.

Yup.

> is that I wish you separated all those commands into multiple files, instead of listing everything in the plugins.js file. Similar to the pattern invoke and invokeLocal. Splitting this up would make it easier to maintain.

It isn't too late :+1: plugin commands are groupings of functionality, so I suggest the following structure. How do you like it?

``` yaml
lib/plugins/

plugin
    ├── install - index.js
    ├── uninstall - index.js
    ├── list - index.js
    └── search - index.js
```

>Imo the --version option should be dropped, and we should instead parse the --name option to get the version, which could look something like abc@1.0.0 since this is a much known pattern and easier than providing a second option.

I like this idea. We should adopt a well-known pattern. will do so.

>install/uninstall commands only run inside a service, which is understandable because you need to read the serverless.yml file. But that means a user can't install a plugin globally? and maybe use the same plugin in multiple services. Don't think it's a priority though.

Yup. Right now, you can't install a plugin globally. 
However, AFAIK I can’t think of any use cases which need to install globally on using the framework. I think had better node modules, including sls plugin, would be managed within own services.

If there are such use cases, we should allow users to globally install via plugin commands, something like adding `--global` flag in the future.

@HyperBrain
Got it, thx!
I will add that after your PR would be merged.

horike37 commented:
@pmuens @eahefnawy @brianneisler 

I'm working on this based on the feedback. However I have faced the two issues and paused in my task.
I'd like to talk about these with you.

The first one is, right now the top level command of `serverless plugin` does not specify any hooks since all operations of the plugin is given by the second level like `serverless plugin list`.
I think the best way is that only the top level `serverless plugin` should be dropped. Is it possible by adding any syntax into where new commands describe?

If not, the top level should display the plugin command help.

The second one is, when installing, running the check whether to be already installed. However, this process is complex a little bit. If you specify a latest as a version(`serverless install -n serverless-webpack@latest`), need to send a request to npm registry so that getting a real version number of the latest.

I think had better drop the check and only run `npm install --save-dev <pluginname>`. The check would be slowing down the command because of the extra http request.


The consistency of installation can be kept only npm w/o the check.

pmuens commented:
Hey @horike37 thanks for updating and thanks for your comments 👍

> I re-thought above for a while. as in conclusion, I have changed my thought that it had better the top level should display the plugin command help.

That sounds like a good solution 👍

> The second one is, when installing, currently, running the check whether to be already installed. However, this process is complex a little bit. If you specify a latest as a version(serverless install -n serverless-webpack@latest), need to send a request to npm registry so that getting a real version number of the latest.
> I think had better drop the check and only run npm install --save-dev <pluginname>. The check would be slowing down the command because of the extra http request.
> The consistency of installation can be kept only npm w/o the check.

So this only affects `@latest`? I'm all in for dropping it for now. IMHO it's better to get a MVP version out (which we can ship in the next release) and then iterate on it (and add this functionality later on).

Good job on the PR so far! Let us know if you need any other help here 💪

horike37 commented:
>So this only affects @latest? 

Yup

>IMHO it's better to get a MVP version out (which we can ship in the next release) and then iterate on it (and add this functionality later on).

yeah, I like this idea :smile:
Thank you for your advice @pmuens!  will continue to do.

horike37 commented:
@pmuens @eahefnawy 
I have fixed all the things which were pointed out. 👍
Please review again.

## The list of what I have done after the second review
- Added the functionality which displays command help to the top level.
- Dropped `--version` option from the install. Instead, updated to get the version from `--name` option like `@<version>`.
- Updated code for turning into async call.
- Dropped the check whether to be already installed or uninstalled.
- Fixed typo for openwhisk doc.
- Dropped current command aliases (add this after merging #4198 later on)

pmuens commented on lib/plugins/install/install.js:
```diff
@@ -15,7 +15,7 @@ class Install {
     this.commands = {
       install: {
-        usage: 'Install a Serverless service from GitHub',
+        usage: 'Install a Serverless service from GitHub or plugin registry',
```
"or a plugin from the Serverless registry"

pmuens commented on lib/plugins/plugin/install/install.js:
```diff
@@ -0,0 +1,158 @@
+          .then(() => {
+            const message = [
+              'Successfully installed',
+              `"${this.options.pluginName}@${this.options.pluginVersion}"`,
```
A space should be added at the beginning of this line.

pmuens commented on lib/plugins/plugin/lib/pluginService.js:
```diff
@@ -0,0 +1,68 @@
+      message = message.slice(0, -2);
+      this.serverless.cli.consoleLog(message);
+      this.serverless.cli.consoleLog(`
+To install a plugin run 'sls plugin install --name plugin-name-here'
```
"sls plugin install" --> "serverless plugin install"

pmuens commented on lib/plugins/plugin/lib/pluginService.js:
```diff
@@ -0,0 +1,68 @@
+      this.serverless.cli.consoleLog(`
+To install a plugin run 'sls plugin install --name plugin-name-here'
+
+It will be automatically downloaded, added to package.json & added to serverless.yml
```
"It will be automatically downloaded and added to your package.json and serverless.yml file"

pmuens commented on lib/plugins/plugin/uninstall/uninstall.test.js:
```diff
@@ -0,0 +1,401 @@
+chai.use(require('chai-as-promised'));
+const expect = require('chai').expect;
+
+describe('pluginUninstall', () => {
```
`PluginUninstall`

pmuens commented on lib/plugins/plugin/lib/pluginService.js:
```diff
@@ -0,0 +1,68 @@
+'use strict';
```
Is `pluginService` meant to be a place for common functionality? If so we should rename it to `utils.js` or something similar so that it's obvious when someone looks through the code.

horike37 replied:
>Is pluginService meant to be a place for common functionality? 
Yup, will fix it.

pmuens commented on lib/plugins/plugin/lib/pluginService.test.js:
```diff
@@ -0,0 +1,155 @@
+chai.use(require('chai-as-promised'));
+const expect = require('chai').expect;
+
+describe('PluginService', () => {
```
This needs to be updated if we change the name of the file (see comment above).

horike37 commented:
Thank you for the review @pmuens :muscle:
Just fixed all on the latest commit. Please check it :wink:

pmuens commented on lib/plugins/plugin/plugin.test.js:
```diff
@@ -0,0 +1,53 @@
+
+    it('should run promise chain in order for "plugin:plugin" hook', () => {
+      const generateCommandsHelpStub = sinon
+        .stub(plugin.serverless.cli, 'generateCommandsHelp').returns(BbPromise.resolve());
```
This should be moved to a `beforeEach` block.

pmuens commented on lib/plugins/plugin/plugin.test.js:
```diff
@@ -0,0 +1,53 @@
+
+      return expect(plugin.hooks['plugin:plugin']()).to.be.fulfilled.then(() => {
+        expect(generateCommandsHelpStub.calledOnce).to.equal(true);
+        plugin.serverless.cli.generateCommandsHelp.restore();
```
This should be moved to an `afterEach` block (so that the stub correctly restored even when the test fails).

pmuens commented on lib/plugins/plugin/uninstall/uninstall.test.js:
```diff
@@ -0,0 +1,401 @@
+      .to.be.fulfilled.then(() => {
+        expect(uninstallStub.calledOnce).to.equal(true);
+
+        pluginUninstall.uninstall.restore();
```
This should be moved into an `afterEach` block.

pmuens commented on lib/plugins/plugin/uninstall/uninstall.test.js:
```diff
@@ -0,0 +1,401 @@
+    });
+
+    it('should run promise chain in order for "plugin:uninstall:uninstall" hook', () => {
+      const uninstallStub = sinon
```
This should be moved into a `beforeEach` block.

pmuens commented on lib/plugins/plugin/install/install.test.js:
```diff
@@ -0,0 +1,419 @@
+      return pluginInstall.hooks['plugin:install:install']().then(() => {
+        expect(installStub.calledOnce).to.equal(true);
+
+        pluginInstall.install.restore();
```
This should be moved into an `afterEach` block.

pmuens commented on lib/plugins/plugin/install/install.test.js:
```diff
@@ -0,0 +1,419 @@
+    });
+
+    it('should run promise chain in order for "plugin:install:install" hook', () => {
+      const installStub = sinon
```
This should be moved into a `beforeEach` block.

pmuens commented on lib/plugins/plugin/list/list.test.js:
```diff
@@ -0,0 +1,78 @@
+      return expect(pluginList.hooks['plugin:list:list']()).to.be.fulfilled.then(() => {
+        expect(listStub.calledOnce).to.equal(true);
+
+        pluginList.list.restore();
```
This should be moved into an `afterEach` block.

pmuens commented on lib/plugins/plugin/list/list.test.js:
```diff
@@ -0,0 +1,78 @@
+    });
+
+    it('should run promise chain in order for "plugin:list:list" hook', () => {
+      const listStub = sinon
```
This should be moved into a `beforeEach` block.

pmuens commented on lib/plugins/plugin/search/search.test.js:
```diff
@@ -0,0 +1,102 @@
+      return expect(pluginSearch.hooks['plugin:search:search']()).to.be.fulfilled.then(() => {
+        expect(searchStub.calledOnce).to.equal(true);
+
+        pluginSearch.search.restore();
```
This should be moved into an `afterEach` block.

pmuens commented on lib/plugins/plugin/search/search.test.js:
```diff
@@ -0,0 +1,102 @@
+    });
+
+    it('should run promise chain in order for "plugin:search:search" hook', () => {
+      const searchStub = sinon
```
This should be moved into a `beforeEach` block.

horike37 commented:
@eahefnawy 
🙌🙌🙌

horike37 commented:
@pmuens 

>We've done similar things in the past which caused some tests to randomly fail / pass. It would be >nice if we could update this here as well, so that the tests are stable 💯.

That makes sense! I didn't notice the case.
(I will keep it in mind after implementing. Good to know :thumbsup:)

just fixed.

>Additionally it would be nice if the PR could be rebased since spotinst functions support just landed this morning in master and it would be great if we could add the Plugin docs in their documentation as well.

Just added.

>However running serverless help after errors out with:

Umm :dizzy_face: This should have been fixed in the previous commit.
I created a fresh project again and tried to reproduce it, but it couldn't on my end
Here is the log.
<img width="815" alt="serverless-webpack" src="https://user-images.githubusercontent.com/1301012/30140225-0f5e7650-93ae-11e7-98a2-884386bfb433.png">

There might be some cases which can reproduce. 
Could you please share the step to reproduce? Does the error occur on your end when runnning the install every time?

horike37 commented:
>Could you please add the list with the plugin commands to docs/providers/spotinst/README.md?

Oh, sorry I overlooked it :sweat_smile: I will add it

>I tried to install serverless-webpack into a service I created from scratch. Unfortunately I errored >out. Here are the logs:

Umm 🤔. 

>I use node --version v8.1.0 and npm --version 5.0.3 on Linux (in a Docker container)

Tnank you for your env info! I will take a deep dive into it tomorrow :thumbsup:

horike37 commented:
@pmuens
Could you please share the docker-compose.yml you use? 
I'd like to build the same env.

pmuens commented:
> @pmuens Could you please share the docker-compose.yml you use? I'd like to build the same env.

Thanks for taking the time to look into this @horike37 👍

I've used the `docker-compose.yml` in the Serverless repository. I ran `docker-compose run serverless-node bash` to attach interactively to the container.

horike37 commented:
@pmuens 
I reproduced it with using node v8.1.0 and npm v5.0.3.
Then, this has been resolved :star::star::star:

The cause was in the install process. I described it as follow

1. Run `npm install --save-dev <plugin>`
2. Detect peerDependencies from a package.json of an installed plugin
3. Write detected peerDependencies to a service package.json as devDependencies.
4. Run `npm install` on a service directory

The 3 and 4 process went wrong..
The status of 3 is that a plugin module has already installed, but modules which is described as peerDependencies is not installed yet. At this time, if running `npm install`, `npm` detect not to be installed peerDependencies and stop the process. As a result, these are not installed.

To resolve it, I have updated the behavior as follow
1. Run `npm install --save-dev <plugin>`
2. Detect peerDependencies from a package.json of an installed plugin
3. Install peerDependencies with using `npm install --save-dev <peerDependency plugin>`

Off course,  peerDependencies module will be installed with a range of version which is described in a package.json.
Please try again.

horike37 commented:
@pmuens 
Finally 🎉🎉🎉🎉🎉🎉🎉🎉🎉 I'm so happy!!!
Thank you for taking time to review! And I appreciate the community member's help here!!

pmuens commented:
> Finally 🎉🎉🎉🎉🎉🎉🎉🎉🎉 I'm so happy!!! Thank you for taking time to review! And I appreciate the community member's help here!!

Thanks for your patience @horike37 😬 😅! Super nice that this finally landed in `master` 🎉
